File: linda-core/src/main/java/it/unibo/coordination/tuples/core/OperationType.java
package it.unibo.coordination.tuples.core;

import java.util.EnumSet;

public enum OperationType {
    GET("get"),

    WRITE("out"),
    READ("rd"),
    TAKE("in"),
    ABSENT("no"),

    WRITE_ALL("out_all"),
    READ_ALL("rd_all"),
    TAKE_ALL("take_all"),

    TRY_READ("rdp"),
    TRY_TAKE("inp"),
    TRY_ABSENT("nop");

    private static final EnumSet<OperationType> NOTHING_ACCEPTING = EnumSet.of(GET);
    private static final EnumSet<OperationType> TEMPLATE_ACCEPTING = EnumSet.of(READ, TAKE, TRY_TAKE, TRY_READ, ABSENT, TRY_ABSENT, READ_ALL, TAKE_ALL);
    private static final EnumSet<OperationType> TEMPLATES_ACCEPTING = EnumSet.noneOf(OperationType.class);
    private static final EnumSet<OperationType> TUPLE_ACCEPTING = EnumSet.of(WRITE);
    private static final EnumSet<OperationType> TUPLES_ACCEPTING = EnumSet.of(WRITE_ALL);

    private static final EnumSet<OperationType> TUPLE_RETURNING = EnumSet.of(WRITE, READ, TAKE, TRY_READ, TRY_TAKE, TRY_ABSENT);
    private static final EnumSet<OperationType> TUPLES_RETURNING = EnumSet.of(GET, WRITE_ALL, READ_ALL, TAKE_ALL, TRY_READ, TRY_TAKE, TRY_ABSENT);
    private static final EnumSet<OperationType> TEMPLATE_RETURNING = EnumSet.of(ABSENT, TRY_ABSENT);
    private static final EnumSet<OperationType> TEMPLATES_RETURNING = EnumSet.of(TRY_ABSENT);

    private final String classicName;

    OperationType(String classicName) {
        this.classicName = classicName;
    }

    public static EnumSet<OperationType> nothingAcceptingSet() {
        return NOTHING_ACCEPTING;
    }

    public static boolean isNothingAccepting(OperationType type) {
        return nothingAcceptingSet().contains(type);
    }

    public static EnumSet<OperationType> templateAcceptingSet() {
        return TEMPLATE_ACCEPTING;
    }

    public static boolean isTemplateAccepting(OperationType type) {
        return templateAcceptingSet().contains(type);
    }

    public static EnumSet<OperationType> templatesAcceptingSet() {
        return TEMPLATES_ACCEPTING;
    }

    public static boolean isTemplatesAccepting(OperationType type) {
        return templatesAcceptingSet().contains(type);
    }

    public static EnumSet<OperationType> tupleAcceptingSet() {
        return TUPLE_ACCEPTING;
    }

    public static boolean isTupleAcceptingSet(OperationType type) {
        return tupleAcceptingSet().contains(type);
    }

    public static EnumSet<OperationType> tuplesAcceptingSet() {
        return TUPLES_ACCEPTING;
    }

    public static boolean isTuplesAcceptingSet(OperationType type) {
        return tuplesAcceptingSet().contains(type);
    }

    public static EnumSet<OperationType> tupleReturningSet() {
        return TUPLE_RETURNING;
    }

    public static boolean isTupleReturningSet(OperationType type) {
        return tupleReturningSet().contains(type);
    }

    public static EnumSet<OperationType> tuplesReturningSet() {
        return TUPLES_RETURNING;
    }

    public static boolean isTuplesReturningSet(OperationType type) {
        return tuplesReturningSet().contains(type);
    }

    public static EnumSet<OperationType> templateReturningSet() {
        return TEMPLATE_RETURNING;
    }

    public static boolean isTemplateReturning(OperationType type) {
        return templateReturningSet().contains(type);
    }

    public static EnumSet<OperationType> templatesReturningSet() {
        return TEMPLATES_RETURNING;
    }

    public static boolean isTemplatesReturning(OperationType type) {
        return templatesReturningSet().contains(type);
    }

    public String getClassicName() {
        return classicName;
    }
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/OperationType.kt
package it.unibo.coordination.linda.core

import java.util.*

enum class OperationType {
    GET,
    SET,

    WRITE,
    READ,
    TAKE,
    ABSENT,

    WRITE_ALL,
    READ_ALL,
    TAKE_ALL,

    TRY_READ,
    TRY_TAKE,
    TRY_ABSENT,

    GET_SIZE,
    GET_PENDING_REQUESTS;


    companion object {

        @JvmStatic
        val NOTHING_ACCEPTING: EnumSet<OperationType> = EnumSet.of(GET, GET_SIZE, GET_PENDING_REQUESTS)

        @JvmStatic
        private val TEMPLATE_ACCEPTING = EnumSet.of(READ, TAKE, TRY_TAKE, TRY_READ, ABSENT, TRY_ABSENT, READ_ALL, TAKE_ALL)

        @JvmStatic
        private val TEMPLATES_ACCEPTING = EnumSet.noneOf(OperationType::class.java)

        @JvmStatic
        private val TUPLE_ACCEPTING = EnumSet.of(WRITE)

        @JvmStatic
        private val TUPLES_ACCEPTING = EnumSet.of(WRITE_ALL)

        @JvmStatic
        private val TUPLE_RETURNING = EnumSet.of(WRITE, READ, TAKE, TRY_READ, TRY_TAKE, TRY_ABSENT)

        @JvmStatic
        private val TUPLES_RETURNING = EnumSet.of(GET, WRITE_ALL, READ_ALL, TAKE_ALL, TRY_READ, TRY_TAKE, TRY_ABSENT)

        @JvmStatic
        private val TEMPLATE_RETURNING = EnumSet.of(ABSENT, TRY_ABSENT)

        @JvmStatic
        private val TEMPLATES_RETURNING = EnumSet.of(TRY_ABSENT)

        @JvmStatic
        private val ANY_RETURNING = EnumSet.of(GET_SIZE, GET_PENDING_REQUESTS)

        @JvmStatic
        fun anyReturning(): EnumSet<OperationType> {
            return ANY_RETURNING
        }

        @JvmStatic
        fun nothingAcceptingSet(): EnumSet<OperationType> {
            return NOTHING_ACCEPTING
        }

        @JvmStatic
        fun isNothingAccepting(type: OperationType): Boolean {
            return nothingAcceptingSet().contains(type)
        }

        @JvmStatic
        fun templateAcceptingSet(): EnumSet<OperationType> {
            return TEMPLATE_ACCEPTING
        }

        @JvmStatic
        fun isTemplateAccepting(type: OperationType): Boolean {
            return templateAcceptingSet().contains(type)
        }

        @JvmStatic
        fun templatesAcceptingSet(): EnumSet<OperationType> {
            return TEMPLATES_ACCEPTING
        }

        @JvmStatic
        fun isTemplatesAccepting(type: OperationType): Boolean {
            return templatesAcceptingSet().contains(type)
        }

        @JvmStatic
        fun tupleAcceptingSet(): EnumSet<OperationType> {
            return TUPLE_ACCEPTING
        }

        @JvmStatic
        fun isTupleAcceptingSet(type: OperationType): Boolean {
            return tupleAcceptingSet().contains(type)
        }

        @JvmStatic
        fun tuplesAcceptingSet(): EnumSet<OperationType> {
            return TUPLES_ACCEPTING
        }

        @JvmStatic
        fun isTuplesAcceptingSet(type: OperationType): Boolean {
            return tuplesAcceptingSet().contains(type)
        }

        @JvmStatic
        fun tupleReturningSet(): EnumSet<OperationType> {
            return TUPLE_RETURNING
        }

        @JvmStatic
        fun isTupleReturningSet(type: OperationType): Boolean {
            return tupleReturningSet().contains(type)
        }

        @JvmStatic
        fun tuplesReturningSet(): EnumSet<OperationType> {
            return TUPLES_RETURNING
        }

        @JvmStatic
        fun isTuplesReturningSet(type: OperationType): Boolean {
            return tuplesReturningSet().contains(type)
        }

        @JvmStatic
        fun templateReturningSet(): EnumSet<OperationType> {
            return TEMPLATE_RETURNING
        }

        @JvmStatic
        fun isTemplateReturning(type: OperationType): Boolean {
            return templateReturningSet().contains(type)
        }

        @JvmStatic
        fun templatesReturningSet(): EnumSet<OperationType> {
            return TEMPLATES_RETURNING
        }

        @JvmStatic
        fun isTemplatesReturning(type: OperationType): Boolean {
            return templatesReturningSet().contains(type)
        }

        @JvmStatic
        fun isAnyReturning(type: OperationType): Boolean {
            return anyReturning().contains(type)
        }
    }
}